olm/operator-registry.md

```diff
@@ -0,0 +1,318 @@
+
+---------------------
+
+_**Polling Updates**_
+
+As an OpenShift administrator, I want a way to configure a poll timer to ensure my index of operators is up to date and references all of the latest operator bundles.
+
+- So that I can get updates over time
+- So that I can know the frequency that those updates will appear
+
+Acceptance Criteria:
+When I set the update frequency on an operator index, OLM will poll and ensure that if there is a new version of the index that those changes are pulled and propagated into the cluster.
+
+Internal details:
+Polling process should be reviewed by quay team to ensure we scale on quay (although these indexes should work on ANY container image registry)
+
+### Implementation Details
+
+The goal of this enhancement is to replace the app-registry implementation in a way that is usable by operator developers and operator author developers. To do this we will build an index image that is defined from a set of operator bundle images that are released independently and optimized in a way that can be productized by operator authors and pipeline developers.
+
+This implementation makes the assumption that a separate process is built that generates container images that contain individual operator bundles for each version of an operator. The intent of the index image is to aggregate these images that are out of scope of this enhancement.
+
+The final implementation goal of this enhancement involves this set of steps:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
```

Comment:
Where and how can both the admin and administer say what channels this operator should be added to?

Reply:
The current plan is to include package metadata in the bundle itself. This is specified in the bundle enhancment.

Future plans discussed (but not written down here) including having a separate, catalog-wide set of metadata describing how updates are laid out in a catalog. When such metadata exists, I imagine we will need to add options to this command or provide new commands for inspecting/modifying.

Reply:
It may be nice to link to that here in a "how to build the update graph" section WDYT?

Reply:
@shawn-hurley when you say "that" are you referring to the bundle enhancement? Or are you referring to evan's description of the catalog wide metadata?

I think the former is in a pr that is yet to merge so I would hesitate to link to it, and the latter would need to be written.

Reply:
Somewhere, either the bundle format OR this PR we need to write down how given the input we will develop the graph IMO I think it would help a lot of folks understand what we are planning.